src/ray/object_manager/object_manager.h

```diff
@@ -97,6 +99,72 @@ class ObjectManagerInterface {
   virtual ~ObjectManagerInterface(){};
 };
 
+class PushManager {
+ public:
+  /// Manages rate limiting of outbound object pushes.
+
+  /// Create a push manager.
+  PushManager(int64_t max_chunks_in_flight)
+      : max_chunks_in_flight_(max_chunks_in_flight) {
+    RAY_CHECK(max_chunks_in_flight_ > 0) << max_chunks_in_flight_;
+  };
+
+  /// Start pushing an object subject to max bytes in flight limit.
+  ///
+  /// \param push_id Unique identifier for this push.
+  /// \param num_chunks The total number of chunks to send.
+  /// \param send_chunk_fn This function will be called with args 0...{num_chunks-1}.
+  ///                      The caller promises to call PushManager::OnChunkComplete()
+  ///                      once a call to send_chunk_fn finishes.
+  void StartPush(const UniqueID &push_id, int64_t num_chunks,
+                 std::function<void(int64_t)> send_chunk_fn);
+
+  /// Called every time a chunk completes to trigger additional sends.
+  /// TODO(ekl) maybe we should cancel the entire push on error.
+  void OnChunkComplete();
+
+  /// Return the number of chunks currently in flight. For testing only.
+  int64_t NumChunksInFlight() const { return chunks_in_flight_; };
+
+  /// Return the number of chunks remaining. For testing only.
+  int64_t NumChunksRemaining() const { return chunks_remaining_; };
+
+  /// Return the number of pushes currently in flight. For testing only.
+  int64_t NumPushesInFlight() const { return push_info_.size(); };
+
+  std::string DebugString() const {
+    std::stringstream result;
+    result << "PushManager:";
+    result << "\n- num pushes in flight: " << NumPushesInFlight();
+    result << "\n- num chunks in flight: " << NumChunksInFlight();
+    result << "\n- num chunks remaining: " << NumChunksRemaining();
+    result << "\n- max chunks allowed: " << max_chunks_in_flight_;
+    return result.str();
+  }
+
+ private:
+  /// Called on completion events to trigger additional pushes.
+  void ScheduleRemainingPushes();
+
+  /// Pair of (num_chunks, chunk_send_fn).
+  typedef std::pair<int64_t, std::function<void(int64_t)>> PushInfo;
+
+  /// Max number of chunks in flight allowed.
+  const int64_t max_chunks_in_flight_;
+
+  /// Running count of chunks remaining.
+  int64_t chunks_remaining_ = 0;
```

Comment:
Could you clarify this description? At first I thought this was the number of chunks that could still be sent (max - in-flight).